src/ur5_simulation/ur5_kinematics.py

```diff
@@ -20,55 +21,150 @@ def __init__(self, d1=0.089159, a2=-0.42500, a3=-0.39225, d4=0.1091,
+            tf = np.dot(tf, self._gen_transf(self.dh_matrix.iloc[link]))
         
         if short == True:
             tf = tf.round(3) 
         
         return tf 
-    
\ No newline at end of file
+
+    ##############  INVERSE KINEMATICS CALCULATION  ############## 
+
+    def _calculate_teta_1(self, tf, shoulder='left'):
+        p05 = np.dot(tf, np.array([[0],[0],[-self.__d6],[1]])) # origin 5 in respect to 1
+        p05_x = p05[0][0]
+        p05_y = p05[1][0]
+        p05_xy = (p05_x**2 + p05_y**2)**0.5
+        phi_1 = atan2(p05_y,p05_x)
+        phi_2 = acos(self.__d4/p05_xy)
+        self._teta1 = phi_1 + phi_2 + pi/2
+        
+        if shoulder == 'right':
+            self._teta1 = phi_1 - phi_2 + pi/2
+
+        return self._teta1
+        
+    def _calculate_teta_5(self, tf, wrist='up'):
+        p06_x = tf[0][3]
+        p06_y = tf[1][3]
+        self._teta5 = acos((p06_x*sin(self._teta1) - p06_y*cos(self._teta1) - self.__d4)/self.__d6)
+
+        if wrist == 'down':
+            self._teta5 *= -1 
+
+        return self._teta5
+
+    def _calculate_teta_6(self, tf):
+        t60 = inv(tf) # transformation matrix from 6 to 0
+        y = (-t60[1][0]*sin(self._teta1) + t60[1][1]*cos(self._teta1))/sin(self._teta5)
+        x = (t60[0][0]*sin(self._teta1) - t60[0][1]*cos(self._teta1))/sin(self._teta5)
+        self._teta6 = atan2(y,x)
+
+        return self._teta6   
+
+    def _calculate_teta_3(self, tf, elbow ='up'):
+        t01 = self._gen_transf(self.dh_matrix.iloc[0]) # transformation matrix from 0 to 1
+        t45 = self._gen_transf(self.dh_matrix.iloc[4]) # transformation matrix from 4 to 5
+        t56 = self._gen_transf(self.dh_matrix.iloc[5]) # transformation matrix from 5 to 6
+
+        # t14 = tf @ inv(t01) @ inv(t45) @ inv(t56) error
+        # t14 = tf @ inv(t01) @ inv(np.dot(t45,t56)) error
+        # t14 = inv(t01) @ tf @ inv(t45) @ inv(t56) error
+        # Conclusion: the order of multiplication matters!!
+
+        t14 = inv(t01) @ tf @ inv(np.dot(t45,t56)) # transformation matrix from 1 to 4
+        self._p14_x = t14[0][3]
+        self._p14_z = t14[2][3]
+        self._p14_xz = (self._p14_x**2 + self._p14_z**2)**0.5
+        self._teta3 = acos((self._p14_xz**2 - self.__a2**2 - self.__a3**2)/(2*self.__a2*self.__a3))
+
+        if elbow == 'down':
+            self._teta3 *= -1 
+
+        return self._teta3
+
+    def _calculate_teta_2(self, tf):
+        
+        phi_1 = atan2(-self._p14_z, -self._p14_x)
+        phi_2 = asin((-self.__a3*sin(self._teta3))/self._p14_xz)
+        self._teta2 = phi_1 - phi_2
+
+        return self._teta2
+
+    def _calculate_teta_4(self, tf):
+        t34 = self._gen_transf(self.dh_matrix.iloc[3]) # transformation matrix from 3 to 4
+        self._teta4 = atan2(t34[1][0], t34[0][0])
+
+        return self._teta4
+
+    def calculate_inverse_kinematics(self, tf, shoulder='left', wrist='up', elbow='up', short=True):
```

Comment:
How do I know where is shoulder, wrist and elbow in the moment of inverse kinematics? They always will be in this position left, up and up respectively?

Reply:
These are geometric factors that depend on how the robot was modeled. For the way the DH matrix was set, the sequence will always be left, up, up. If the user inputs shoulder='right', for example, the links length and orientation need to be reassessed. Same is valid for elbow and wrist.

Reply:
I have studied the Spong Book to understand better this "problem" and the position of components doesn't depend on how the arm was modeled but how is the configuration to the arm end-effector reaches some pose. Below it's a good example. If you have Spong book first edition, see figures 3.17, 3.18, and 3.20. But anyway, I have sure that all positions chosen to validate the kinematics methods have the arm configuration with `shoulder = left` and `elbow = up`. But I think `wrist` varies between up and down. Please, see the [pictures](https://github.com/matosinho/enga73_latex/tree/dev-mateus/images) in our document to see the end effector poses which we are using.